code/modules/xenoarchaeology/xenoartifact.dm

```diff
@@ -0,0 +1,470 @@
+	. = ..()
+	if(process_type == PROCESS_TYPE_LIT) //Snuff out candle
+		to_chat(user, "<span class='notice'>You snuff out [name]</span>")
+		process_type = null
+		return
+	if(user.a_intent == INTENT_GRAB)
+		if(touch_desc?.on_touch(src, user) && user.can_see_reagents())
+			balloon_alert(user, (initial(touch_desc.desc) ? initial(touch_desc.desc) : initial(touch_desc.label_name)), material)
+		return
+	SEND_SIGNAL(src, XENOA_INTERACT, null, user, user)
+
+/obj/item/xenoartifact/attackby(obj/item/I, mob/living/user, params)
+	var/tool_text
+	for(var/datum/xenoartifact_trait/t as() in traits) //chat & bubble hints & helpers
+		if(t?.on_item(src, user, I) && user.can_see_reagents())
+			tool_text = "[tool_text][t.desc ? t.desc : t.label_name]\n"
+	if(tool_text)
+		balloon_alert(user, tool_text, material)
+
+	//abort if grab intent
+	if(!(COOLDOWN_FINISHED(src, xenoa_cooldown))||user?.a_intent == INTENT_GRAB||istype(I, /obj/item/xenoartifact_label)||istype(I, /obj/item/xenoartifact_labeler))
+		if(user?.a_intent == INTENT_GRAB)
+			to_chat(user, "<span class='notice'>You preform a safe operation on [src] with [I].</span>")
+		return
+	else if(istype(I, /obj/item/wirecutters) && (locate(/obj/item/xenoartifact_label) in contents)) //allow people to remove stickers
+		label_desc = null
+		qdel(locate(/obj/item/xenoartifact_label) in contents)
+	..()
+
+///Run traits. Used to activate all minor, major, and malfunctioning traits in the artifact's trait list. Sets cooldown when properly finished.
+/obj/item/xenoartifact/proc/check_charge(mob/user, charge_mod)
+	log_game("[user] attempted to activate [src] at [world.time]. Located at [x] [y] [z].")
+
+	if(COOLDOWN_FINISHED(src, xenoa_cooldown))
+		if(prob(malfunction_chance) && traits.len < 6 + (material == XENOA_URANIUM ? 1 : 0)) //See if we pick up an malfunction
+			generate_trait_unique(XENOA_MALFS)
+			malfunction_chance = 0 //Lower chance after contracting 
+		else //otherwise increase chance.
+			//Ramps malf_mod down if it's going to create a value > 100
+			malfunction_chance += (malfunction_chance+malfunction_mod < 100 ? malfunction_mod : malfunction_mod-((malfunction_chance+malfunction_mod)-100))
+
+		charge += charge_mod
+		for(var/datum/xenoartifact_trait/minor/t in traits)//Minor traits aren't apart of the target loop
+			t?.activate(src, user, user)
+			log_game("[src] activated minor trait [t] at [world.time]. Located at [x] [y] [z]")
+		charge = (charge+charge_req)/1.9 //Not quite an average. Generally produces better results.   
+		for(var/atom/M in true_target) //target loop
+			if(get_dist(get_turf(src), get_turf(M)) <= max_range) 
+				create_beam(M) //Indicator beam, points to target, M
+				for(var/datum/xenoartifact_trait/t as() in traits) //Major traits
+					if(istype(t, /datum/xenoartifact_trait/major) || istype(t, /datum/xenoartifact_trait/malfunction))
+						log_game("[src] activated major trait [t] at [world.time]. Located at [x] [y] [z]")
+						t.activate(src, M, user)
```

Comment:
This code might be cleaner if you had a parameter indicating that its a major activation on t.activate and then ignoring non-major activations on /datum/xenoartifact_trait/minor in an overriden version of the proc.